math/int.go

```diff
@@ -420,10 +421,90 @@ func (i *Int) Unmarshal(data []byte) error {
 	return nil
 }
 
+var bigMaxUint64 = new(big.Int).SetUint64(uint64(stdmath.MaxUint64))
+
 // Size implements the gogo proto custom type interface.
-func (i *Int) Size() int {
-	bz, _ := i.Marshal()
-	return len(bz)
+// Reduction power of 10 is the smallest power of 10, than 1<<64-1
+//
+//	18446744073709551615
+//
+// and the next value fitting with the digits of (1<<64)-1 is:
+//
+//	10000000000000000000
+var big10Pow19, _ = new(big.Int).SetString("1"+strings.Repeat("0", 19), 10)
+var log10Of2 = stdmath.Log10(2)
+
+func (i *Int) Size() (size int) {
+	sign := i.Sign()
+	if sign == 0 { // It is zero.
+		// log*(0) is undefined hence return early.
+		return 1
+	}
+
+	ii := i.i
+	alreadyMadeCopy := false
+	if sign < 0 { // Negative sign encountered, so consider len("-")
+		// The reason that we make this comparison in here is to
+		// allow checking for negatives exactly once, to reduce
+		// on comparisons inside sizeBigInt, hence we make a copy
+		// of ii and make it absolute having taken note of the sign
+		// already.
+		size += 1
+		// We already accounted for the negative sign above, thus
+		// we can now compute the length of the absolute value.
+		ii = new(big.Int).Abs(ii)
+		alreadyMadeCopy = true
+	}
+
+	// From here on, we are now dealing with non-0, non-negative values.
+	return size + sizeBigInt(ii, alreadyMadeCopy)
+}
+
+func sizeBigInt(i *big.Int, alreadyMadeCopy bool) (size int) {
+	// This code assumes that non-0, non-negative values have been passed in.
+	bitLen := i.BitLen()
+	res := float64(bitLen) * log10Of2
+	ires := int(res)
+	if diff := res - float64(ires); diff == 0.0 {
+		return size + ires
+	} else if diff >= 0.3 { // There are other digits past the bitLen, this is a heuristic.
+		return size + ires + 1
+	}
+
+	// Use Log10(x) for values less than (1<<64)-1, given it is only defined for [1, (1<<64)-1]
+	if i.Cmp(bigMaxUint64) <= 0 {
+		return size + 1 + int(stdmath.Log10(float64(i.Uint64())))
+	}
+
+	// Past this point, the value is greater than (1<<64)-1 and 10^19.
+
+	// The prior above computation of i.BitLen() * log10Of2 is inaccurate for powers of 10
+	// and values like "9999999999999999999999999999"; that computation always overshoots by 1
+	// hence our next alternative is to just go old school and keep dividing the value by:
+	//   10^19 aka "10000000000000000000" while incrementing size += 19
+
+	// At this point we should just keep reducing by 10^19 as that's the smallest multiple
+	// of 10 that matches the digit length of (1<<64)-1
+	var ri *big.Int
```

Comment:
Can we pass this in as a scratch variable each loop, to avoid re-allocations?

Reply:
ping @odeke-em